File: src/main/java/io/github/bobisawesome07/world/dimension/ModDimensions.java
package io.github.bobisawesome07.world.dimension;

import io.github.bobisawesome07.FracturedMod;
import io.github.bobisawesome07.world.BarrierFloorChunkGenerator;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.registry.Registerable;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.registry.tag.BlockTags;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.intprovider.UniformIntProvider;
import net.minecraft.world.Difficulty;
import net.minecraft.world.GameRules;
import net.minecraft.world.World;
import net.minecraft.world.biome.BiomeKeys;
import net.minecraft.world.dimension.DimensionType;
import net.minecraft.world.dimension.DimensionTypes;
import net.minecraft.world.gen.chunk.ChunkGeneratorSettings;
import xyz.nucleoid.fantasy.Fantasy;
import xyz.nucleoid.fantasy.RuntimeWorldConfig;
import xyz.nucleoid.fantasy.RuntimeWorldHandle;

import java.util.Map;
import java.util.OptionalLong;
import java.util.concurrent.ConcurrentHashMap;

import static io.github.bobisawesome07.FracturedMod.MOD_ID;
import static io.github.bobisawesome07.FracturedMod.server;

public class ModDimensions {
    // Registry key for the pocket dimension type
    public static final RegistryKey<DimensionType> POCKET_DIM_TYPE = RegistryKey.of(
            RegistryKeys.DIMENSION_TYPE,
            new Identifier(MOD_ID, "pocket_dimension")
    );

    // Fantasy instance for managing runtime worlds
    private static final Fantasy fantasy = Fantasy.get(FracturedMod.getServer());

    // Cache to hold created pocket dimensions
    private static final Map<String, RuntimeWorldHandle> worldCache = new ConcurrentHashMap<>();

    // Chunk generator for pocket dimensions with a barrier floor
    private static final BarrierFloorChunkGenerator pocketGen = new BarrierFloorChunkGenerator(
            FracturedMod.getServer().getRegistryManager()
                    .get(RegistryKeys.BIOME)
                    .entryOf(BiomeKeys.THE_END)
    );

    // Configuration for pocket dimensions
    private static final RuntimeWorldConfig worldConfig = new RuntimeWorldConfig()
            .setDimensionType(POCKET_DIM_TYPE)
            .setDifficulty(Difficulty.HARD)
            .setGameRule(GameRules.DO_DAYLIGHT_CYCLE, false)
            .setGenerator(pocketGen)
            .setSeed(1234L);

    /**
     * Teleports a player to their pocket dimension
     * <p>
     * * @param world Source world
     * * @param user Player to teleport
     */
    public static void tpToPocket(World world, PlayerEntity user) {
        ServerPlayerEntity serverPlayer = (ServerPlayerEntity) user;
        ServerWorld targetWorld = server.getWorld(RegistryKey.of(
                RegistryKeys.WORLD,
                new Identifier(MOD_ID, "pocket_dimension" + user.getUuidAsString())
        ));

        if (targetWorld != null) {
            serverPlayer.teleport(targetWorld, 0.0, 2.0, 0.0, 0f, 0f);
        }
    }

    /**
     * Creates a new pocket dimension or loads an existing one
     * <p>
     * * @param nameSpace Namespace for the dimension identifier
     * * @param uuid Player UUID to create a unique dimension ID
     * * @return Handle to the runtime world
     */
    public static RuntimeWorldHandle createOrLoadPocketDimension(String nameSpace, String uuid) {
        String worldId = nameSpace + ":pocket_dimension" + uuid;
        return worldCache.computeIfAbsent(worldId, id ->
                fantasy.getOrOpenPersistentWorld(
                        new Identifier(nameSpace, "pocket_dimension" + uuid),
                        worldConfig
                )
        );
    }

    /**
     * Registers the pocket dimension type during game bootstrap
     * <p>
     * * @param context Registry context
     */
    public static void bootstrapType(Registerable<DimensionType> context) {
        context.register(POCKET_DIM_TYPE, new DimensionType(
                OptionalLong.of(0),  // Fixed time
                false,              // Skylight
                false,              // Ceiling
                false,              // Ultra warm
                false,              // Natural
                1.0,                // Coordinate scale
                false,              // Bed works
                false,              // Respawn anchor works
                0,                  // Min Y
                256,                // Height
                256,                // Logical height
                BlockTags.INFINIBURN_OVERWORLD,
                DimensionTypes.THE_END_ID,
                1.0f,
                new DimensionType.MonsterSettings(false, false, UniformIntProvider.create(0, 0), 0)
        ));
    }
}

File: src/main/java/io/github/bobisawesome07/world/dimension/ModDimensions.kt
package io.github.bobisawesome07.world.dimension

import io.github.bobisawesome07.FracturedMod
import io.github.bobisawesome07.world.BarrierFloorChunkGenerator
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.registry.Registerable
import net.minecraft.registry.RegistryKey
import net.minecraft.registry.RegistryKeys
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import net.minecraft.world.Difficulty
import net.minecraft.world.GameRules
import net.minecraft.world.World
import net.minecraft.world.biome.BiomeKeys
import net.minecraft.world.dimension.DimensionType
import xyz.nucleoid.fantasy.Fantasy
import xyz.nucleoid.fantasy.RuntimeWorldConfig
import xyz.nucleoid.fantasy.RuntimeWorldHandle
import java.util.concurrent.ConcurrentHashMap

object ModDimensions {
    // Registry key for the pocket dimension type
    val POCKET_DIM_TYPE: RegistryKey<DimensionType> = RegistryKey.of(
        RegistryKeys.DIMENSION_TYPE,
        Identifier(FracturedMod.MOD_ID, "pocket_dimension")
    )

    // Fantasy instance for managing runtime worlds
    private val fantasy: Fantasy = Fantasy.get(FracturedMod.server)

    // Cache to hold created pocket dimensions
    private val worldCache: MutableMap<String, RuntimeWorldHandle> = ConcurrentHashMap()

    // Cache to hold entry locations
    private val entryLocationCache: MutableMap<String, BlockPos> = ConcurrentHashMap()

    // Chunk generator for pocket dimensions with a barrier floor
    private val pocketGen = BarrierFloorChunkGenerator(
        FracturedMod.server!!.registryManager
            .get(RegistryKeys.BIOME)
            .entryOf(BiomeKeys.BEACH)
    )

    // Configuration for pocket dimensions
    private val worldConfig: RuntimeWorldConfig = RuntimeWorldConfig()
        .setDimensionType(POCKET_DIM_TYPE)
        .setDifficulty(Difficulty.HARD)
        .setGameRule(GameRules.DO_DAYLIGHT_CYCLE, false)
        .setGenerator(pocketGen)
        .setSeed(1234L)

    /**
     * Teleports the specified player to their personal pocket dimension and records their entry location.
     *
     * Stores the player's current block position before teleporting them to coordinates (0, 2, 0) in their unique pocket dimension.
     *
     * @param world The world from which the player is being teleported.
     * @param user The player to teleport.
     */
    @JvmStatic
    fun tpToPocket(world: World?, user: PlayerEntity) {
        val serverPlayer = user as ServerPlayerEntity
        val targetWorld = FracturedMod.server!!.getWorld(
            RegistryKey.of(
                RegistryKeys.WORLD,
                Identifier(FracturedMod.MOD_ID, "pocket_dimension" + user.getUuidAsString())
            )
        )

        if (targetWorld != null) {
            // Store the entry location
            entryLocationCache[user.uuidAsString] = user.blockPos

            serverPlayer.teleport(targetWorld, 0.0, 2.0, 0.0, 0f, 0f)
        }
    }

    /**
     * Teleports the player back to their previously stored entry location in the specified world.
     *
     * If an entry location was saved for the player, they are moved to that position with zero yaw and pitch.
     *
     * @param world The world to teleport the player into.
     * @param user The player to teleport.
     */
    @JvmStatic
    fun tpBackToEntry(world: World?, user: PlayerEntity) {
        val serverPlayer = user as ServerPlayerEntity
        val entryLocation = entryLocationCache[user.uuidAsString]

        if (entryLocation != null) {
            serverPlayer.teleport(world, entryLocation.x.toDouble(), entryLocation.y.toDouble(), entryLocation.z.toDouble(), 0f, 0f)
        }
    }

    /**
     * Creates a new pocket dimension or loads an existing one
     *
     * @param nameSpace Namespace for the dimension identifier
     * @param uuid Player UUID to create a unique dimension ID
     * @return Handle to the runtime world
     */
    @JvmStatic
    fun createOrLoadPocketDimension(nameSpace: String, uuid: String): RuntimeWorldHandle {
        val worldId = "$nameSpace:pocket_dimension$uuid"
        return worldCache.computeIfAbsent(worldId) { id: String? ->
            fantasy.getOrOpenPersistentWorld(
                Identifier(nameSpace, "pocket_dimension$uuid"),
                worldConfig
            )
        }
    }

    /**
     * Registers the pocket dimension type during game bootstrap
     *
     * @param context Registry context
     */
    fun bootstrapType(context: Registerable<DimensionType?>?) {
        register()
    }
}

private fun register() {

}
